File: articles.jdbc/src/test/java/pet/db/jdbc/service/UserServiceTest.java
package pet.db.jdbc.service;


import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import pet.db.jdbc.model.dto.Article;
import pet.db.jdbc.model.dto.User;
import pet.db.jdbc.tool.db.DbCleaner;
import pet.db.jdbc.tool.generator.TestDataGenerator;
import pet.db.jdbc.tool.exception.DuplicateUserException;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertThrows;

@SpringBootTest
public class UserServiceTest {

    @Autowired
    private DbCleaner dbCleaner;

    @Autowired
    private UserService userService;

    @Autowired
    private TestDataGenerator<User> userTestDataGenerator;

    @Autowired
    private TestDataGenerator<Article> articleTestDataGenerator;

    @AfterEach
    void cleanDb() {
        dbCleaner.cleanAll();
    }

    @Test
    public void saveUser() throws DuplicateUserException {
        User userForSave = userTestDataGenerator.generateUnsavedData();

        User savedUser = userService.create(userForSave);

        Optional<User> userOptionalForCheck = userService.findById(savedUser.getId());
        assertTrue(userOptionalForCheck.isPresent());
        assertEquals(savedUser, userOptionalForCheck.get());
    }

    @Test
    public void saveUserWithInvalidData() {
        User userForSave = userTestDataGenerator.generateUnsavedData();
        userForSave.setUsername(null);

        assertThrows(RuntimeException.class, () -> userService.create(userForSave));
    }

    @Test
    public void saveUserWithNotUniqueUsername() {
        User savedUser = userTestDataGenerator.generateSavedData();
        User unsavedUser = userTestDataGenerator.generateUnsavedData();
        unsavedUser.setUsername(savedUser.getUsername());

        assertThrows(DuplicateUserException.class, () -> userService.create(unsavedUser));
    }

    @Test
    public void saveUserWithNotUniqueEmail() {
        User savedUser = userTestDataGenerator.generateSavedData();
        User unsavedUser = userTestDataGenerator.generateUnsavedData();
        unsavedUser.setEmail(savedUser.getEmail());

        assertThrows(DuplicateUserException.class, () -> userService.create(unsavedUser));
    }

    @Test
    public void updateUserById() throws DuplicateUserException {
        User savedUser = userTestDataGenerator.generateSavedData();
        User userDataForUpdate = userTestDataGenerator.generateUnsavedData();

        User updatedUser = userService.updateById(userDataForUpdate, savedUser.getId());

        Optional<User> userOptionalForCheck = userService.findById(savedUser.getId());
        assertTrue(userOptionalForCheck.isPresent());
        assertEquals(updatedUser, userOptionalForCheck.get());
    }

    @Test
    public void updateUserByIdWithSameUsernameAndEmail() throws DuplicateUserException {
        User savedUser = userTestDataGenerator.generateSavedData();
        User userDataForUpdate = userTestDataGenerator.generateUnsavedData();
        userDataForUpdate.setUsername(savedUser.getUsername());
        userDataForUpdate.setEmail(savedUser.getEmail());

        User updatedUser = userService.updateById(userDataForUpdate, savedUser.getId());

        Optional<User> userOptionalForCheck = userService.findById(savedUser.getId());
        assertTrue(userOptionalForCheck.isPresent());
        assertEquals(updatedUser, userOptionalForCheck.get());
    }

    @Test
    public void updateUserByNonExistentId() {
        User unsavedUser = userTestDataGenerator.generateUnsavedData();
        User userDataForUpdate = userTestDataGenerator.generateUnsavedData();

        assertThrows(NoSuchElementException.class,
                () -> userService.updateById(userDataForUpdate, unsavedUser.getId()));
    }

    @Test
    public void updateUserByIdWithNotUniqueUsername() {
        User savedUser = userTestDataGenerator.generateSavedData();
        User userDataForUpdate = userTestDataGenerator.generateUnsavedData();
        User anotherSavedUser = userTestDataGenerator.generateSavedData();
        userDataForUpdate.setUsername(anotherSavedUser.getUsername());

        assertThrows(DuplicateUserException.class,
                () -> userService.updateById(userDataForUpdate, savedUser.getId()));
    }

    @Test
    public void updateUserByIdWithNotUniqueEmail() {
        User savedUser = userTestDataGenerator.generateSavedData();
        User newUserDataForUpdate = userTestDataGenerator.generateUnsavedData();
        User anotherSavedUser = userTestDataGenerator.generateSavedData();
        newUserDataForUpdate.setEmail(anotherSavedUser.getEmail());

        assertThrows(DuplicateUserException.class,
                () -> userService.updateById(newUserDataForUpdate, savedUser.getId()));
    }

    @Test
    public void deleteUserById()  {
        User savedUser = userTestDataGenerator.generateSavedData();

        userService.deleteById(savedUser.getId());

        assertFalse(userService.existsById(savedUser.getId()));
    }

    @Test
    public void deleteUserByNonExistentId()  {
        User unsavedUser = userTestDataGenerator.generateUnsavedData();

        userService.deleteById(unsavedUser.getId());

        assertFalse(userService.existsById(unsavedUser.getId()));
    }

    @Test
    public void findUserById() {
        User savedUser = userTestDataGenerator.generateSavedData();

        Optional<User> userOptionalForCheck = userService.findById(savedUser.getId());

        assertTrue(userOptionalForCheck.isPresent());
        assertEquals(savedUser, userOptionalForCheck.get());
    }

    @Test
    public void findUserByNonExistentId() {
        User unsavedUser = userTestDataGenerator.generateUnsavedData();

        Optional<User> userOptionalForCheck = userService.findById(unsavedUser.getId());

        assertTrue(userOptionalForCheck.isEmpty());
    }

    @Test
    public void findUserByUsername() {
        User savedUser = userTestDataGenerator.generateSavedData();

        Optional<User> userOptionalForCheck = userService.findByUsername(savedUser.getUsername());

        assertTrue(userOptionalForCheck.isPresent());
        assertEquals(savedUser, userOptionalForCheck.get());
    }

    @Test
    public void findUserByNonExistentUsername() {
        User unsavedUser = userTestDataGenerator.generateUnsavedData();

        Optional<User> userOptionalForCheck = userService.findByUsername(unsavedUser.getUsername());

        assertTrue(userOptionalForCheck.isEmpty());
    }

    @Test
    public void findAuthorsByArticleId() {
        Article savedArticle = articleTestDataGenerator.generateSavedData();
        List<User> authors = userService.findAll();

        List<User> authorsForCheck = userService.findAuthorsByArticleId(savedArticle.getId());

        assertEquals(authors.size(), authorsForCheck.size());
        assertTrue(authors.containsAll(authorsForCheck));
        assertTrue(authorsForCheck.containsAll(authors));
    }

    @Test
    public void findAuthorsByNonExistentArticleId() {
        Article unsavedArticle = articleTestDataGenerator.generateUnsavedData();

        assertThrows(NoSuchElementException.class,
                () -> userService.findAuthorsByArticleId(unsavedArticle.getId()));
    }

    @Test
    public void findAllUsers() {
        List<User> allUsers = userTestDataGenerator.generateSavedData(10);

        List<User> usersForCheck = userService.findAll();

        assertEquals(allUsers.size(), usersForCheck.size());
        assertTrue(allUsers.containsAll(usersForCheck));
        assertTrue(usersForCheck.containsAll(allUsers));
    }

}

File: src/test/kotlin/pet/articles/service/UserServiceTest.kt
package pet.articles.service

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.koin.core.qualifier.named
import org.koin.test.KoinTest
import org.koin.test.inject
import pet.articles.model.dto.Article

import pet.articles.model.dto.User
import pet.articles.service.user.UserService
import pet.articles.tool.exception.DuplicateUserException
import pet.articles.tool.generator.TestDataGenerator
import pet.articles.tool.testing.extension.DBCleanupExtension
import pet.articles.tool.testing.extension.KoinConfigureTestExtension

import java.util.NoSuchElementException

@ExtendWith(DBCleanupExtension::class)
@ExtendWith(KoinConfigureTestExtension::class)
class UserServiceTest : KoinTest {

    companion object {
        const val NUM_OF_TEST_USERS = 10
    }


    private val userService: UserService by inject()

    private val userGenerator: TestDataGenerator<User> by inject(
        named("UserTestDataGenerator")
    )

    private val articleGenerator: TestDataGenerator<Article> by inject(
        named("ArticleTestDataGenerator")
    )

    @Test
    fun saveUser() {
        val userForSave: User = userGenerator.generateUnsavedData()

        val savedUser: User = userService.create(userForSave)

        val userForCheck: User? = userService.findById(savedUser.id!!)
        assertNotNull(userForCheck)
        assertEquals(savedUser, userForCheck)
    }

    @Test
    fun saveUserWithInvalidData() {
        val invalidUser: User = userGenerator.generateInvalidData()
        
        assertThrows(RuntimeException::class.java) {
            userService.create(invalidUser)
        }
    }

    @Test
    fun saveUserWithNotUniqueUsername() {
        val savedUser: User = userGenerator.generateSavedData()
        val unsavedUser: User = userGenerator.generateUnsavedData().copy(
            username = savedUser.username
        )

        assertThrows(DuplicateUserException::class.java) {
            userService.create(unsavedUser)
        }
    }

    @Test
    fun saveUserWithNotUniqueEmail() {
        val savedUser: User = userGenerator.generateSavedData()
        val unsavedUser: User = userGenerator.generateUnsavedData().copy(
            email = savedUser.email
        )

        assertThrows(DuplicateUserException::class.java) {
            userService.create(unsavedUser)
        }
    }

    @Test
    fun updateUserById() {
        val savedUser: User = userGenerator.generateSavedData()
        val userDataForUpdate: User = userGenerator.generateUnsavedData()

        val updatedUser: User = userService.updateById(userDataForUpdate, savedUser.id!!)

        val userForCheck: User? = userService.findById(savedUser.id!!)
        assertNotNull(userForCheck)
        assertEquals(updatedUser, userForCheck)
    }

    @Test
    fun updateUserByIdWithInvalidData() {
        val invalidUser: User = userGenerator.generateInvalidData()
        val savedUser: User = userGenerator.generateSavedData()

        assertThrows(RuntimeException::class.java) {
            userService.updateById(invalidUser, savedUser.id!!)
        }
    }

    @Test
    fun updateUserByIdWithSameUsernameAndEmail() {
        val savedUser: User = userGenerator.generateSavedData()
        val userDataForUpdate: User = userGenerator.generateUnsavedData().copy(
            username = savedUser.username,
            email = savedUser.email
        )

        val updatedUser: User = userService.updateById(userDataForUpdate, savedUser.id!!)

        val userForCheck: User? = userService.findById(savedUser.id!!)
        assertNotNull(userForCheck)
        assertEquals(updatedUser, userForCheck)
    }

    @Test
    fun updateUserByNonExistentId() {
        val unsavedUser: User = userGenerator.generateUnsavedData()
        val userDataForUpdate: User = userGenerator.generateUnsavedData()

        assertThrows(NoSuchElementException::class.java) {
            userService.updateById(userDataForUpdate, unsavedUser.id!!)
        }
    }

    @Test
    fun updateUserByIdWithNotUniqueUsername() {
        val savedUser: User = userGenerator.generateSavedData()
        val userDataForUpdate: User = userGenerator.generateUnsavedData()
        val anotherSavedUser: User = userGenerator.generateSavedData()
        val updatedUser: User = userDataForUpdate.copy(
            username = anotherSavedUser.username
        )

        assertThrows(DuplicateUserException::class.java) {
            userService.updateById(updatedUser, savedUser.id!!)
        }
    }

    @Test
    fun updateUserByIdWithNotUniqueEmail() {
        val savedUser: User = userGenerator.generateSavedData()
        val newUserDataForUpdate: User = userGenerator.generateUnsavedData()
        val anotherSavedUser: User = userGenerator.generateSavedData()
        val updatedUser: User = newUserDataForUpdate.copy(
            email = anotherSavedUser.email
        )

        assertThrows(DuplicateUserException::class.java) {
            userService.updateById(updatedUser, savedUser.id!!)
        }
    }

    @Test
    fun deleteUserById() {
        val savedUser: User = userGenerator.generateSavedData()

        userService.deleteById(savedUser.id!!)

        assertFalse(userService.existsById(savedUser.id!!))
    }

    @Test
    fun deleteUserByNonExistentId() {
        val unsavedUser: User = userGenerator.generateUnsavedData()

        userService.deleteById(unsavedUser.id!!)

        assertFalse(userService.existsById(unsavedUser.id!!))
    }

    @Test
    fun findUserById() {
        val savedUser: User = userGenerator.generateSavedData()

        val userForCheck: User? = userService.findById(savedUser.id!!)

        assertNotNull(userForCheck)
        assertEquals(savedUser, userForCheck)
    }

    @Test
    fun findUserByNonExistentId() {
        val unsavedUser: User = userGenerator.generateUnsavedData()

        val userForCheck: User? = userService.findById(unsavedUser.id!!)

        assertNull(userForCheck)
    }

    @Test
    fun findUserByUsername() {
        val savedUser: User = userGenerator.generateSavedData()

        val userForCheck: User? = userService.findByUsername(savedUser.username)

        assertNotNull(userForCheck)
        assertEquals(savedUser, userForCheck)
    }

    @Test
    fun findUserByNonExistentUsername() {
        val unsavedUser: User = userGenerator.generateUnsavedData()

        val userForCheck: User? = userService.findByUsername(unsavedUser.username)

        assertNull(userForCheck)
    }

    @Test
    fun findAuthorsByArticleId() {
        val savedArticle: Article = articleGenerator.generateSavedData()
        val authors: List<User> = userService.findAll()

        val authorsForCheck: List<User> = userService.findAuthorsByArticleId(savedArticle.id!!)

        assertEquals(authors.size, authorsForCheck.size)
        assertTrue(authors.containsAll(authorsForCheck))
        assertTrue(authorsForCheck.containsAll(authors))
    }

    @Test
    fun findAuthorsByNonExistentArticleId() {
        val unsavedArticle: Article = articleGenerator.generateUnsavedData()

        val authors: List<User> = userService.findAuthorsByArticleId(unsavedArticle.id!!)

        assertTrue(authors.isEmpty())
    }

    @Test
    fun findAllUsers() {
        val allUsers: List<User> = userGenerator.generateSavedData(NUM_OF_TEST_USERS)

        val usersForCheck: List<User> = userService.findAll()

        assertEquals(allUsers.size, usersForCheck.size)
        assertTrue(allUsers.containsAll(usersForCheck))
        assertTrue(usersForCheck.containsAll(allUsers))
    }
}